sdk/tests/simulate-invariant-swap.test.ts

```diff
@@ -0,0 +1,608 @@
+      expect(simulation.stateOutdated).to.equal(false)
+      expect(simulation.maxTicksCrossed).to.equal(false)
+
+      expect(swapResult.amountIn).to.equal(simulation.amountIn)
+      expect(swapResult.amountOut).to.equal(simulation.amountOut)
+      expect(swapResult.startSqrtPrice).to.equal(simulation.startSqrtPrice)
+      expect(swapResult.targetSqrtPrice).to.equal(simulation.targetSqrtPrice)
+      expect(swapResult.fee).to.equal(simulation.fee)
+      expect(simulation.crossedTicks.length).to.equal(0)
+    })
+
+    it('Y to X', async () => {
+      const poolKey = newPoolKey(token0Address, token1Address, feeTier)
+      const pool = await invariant.getPool(token0Address, token1Address, feeTier)
+      const ticks = await Promise.all([
+        invariant.getTick(poolKey, 10n),
+        invariant.getTick(poolKey, -10n)
+      ])
+      const tickmap: Tickmap = await invariant.getRawTickmap(poolKey, 0n)
+      const sqrtPriceLimit = getMaxSqrtPrice(feeTier.tickSpacing)
+
+      const amountIn = 4888n
+      const byAmountIn = false
+      const xToY = false
+
+      const simulation = simulateInvariantSwap(
+        tickmap,
+        protocolFee,
+        feeTier,
+        pool,
+        ticks,
+        xToY,
+        amountIn,
+        byAmountIn,
+        sqrtPriceLimit
+      )
+
+      const swap = await invariant.swap(
+        account,
+        poolKey,
+        xToY,
+        amountIn,
+        byAmountIn,
+        sqrtPriceLimit
+      )
+      expect(swap.events.length).to.equal(1)
+      const swapResult = swap.events[0] as SwapEvent
+
+      expect(simulation.globalInsufficientLiquidity).to.equal(false)
+      expect(simulation.stateOutdated).to.equal(false)
+      expect(simulation.maxTicksCrossed).to.equal(false)
+
+      expect(swapResult.amountIn).to.equal(simulation.amountIn)
+      expect(swapResult.amountOut).to.equal(simulation.amountOut)
+      expect(swapResult.startSqrtPrice).to.equal(simulation.startSqrtPrice)
+      expect(swapResult.targetSqrtPrice).to.equal(simulation.targetSqrtPrice)
+      expect(swapResult.fee).to.equal(simulation.fee)
+      expect(simulation.crossedTicks.length).to.equal(0)
+    })
+
+    it('X to Y', async () => {
+      const poolKey = newPoolKey(token0Address, token1Address, feeTier)
+      const pool = await invariant.getPool(token0Address, token1Address, feeTier)
+      const ticks = await Promise.all([
+        invariant.getTick(poolKey, 10n),
+        invariant.getTick(poolKey, -10n)
+      ])
+      const tickmap: Tickmap = await invariant.getRawTickmap(poolKey, 0n)
+      const sqrtPriceLimit = getMinSqrtPrice(feeTier.tickSpacing)
+
+      const amountIn = 4888n
+      const byAmountIn = false
+      const xToY = true
+
+      const simulation = simulateInvariantSwap(
+        tickmap,
+        protocolFee,
+        feeTier,
+        pool,
+        ticks,
+        xToY,
+        amountIn,
+        byAmountIn,
+        sqrtPriceLimit
+      ) as CalculateSwapResult
```

Comment:
Maybe it's worth adding a wrapping function here to enforce type checking.
For example, this casting could be excluded:
```Typescript
as CalculateSwapResult
```